Initial webservice

bradtaylor opened the pull request:
This is a large PR representing the bulk of the work setting up the service, including the routing, authentication, and notebook conversion. The readme has info on the choice of web framework.

There is still some housekeeping to be done- adding docstrings to the utility methods and creating API documentation.

I also need to create the app configuration for deploying the service on Google App Engine.

Commit:
Added initial server
Set up a small sanic app with two routes:
1. A status route that returns 200 and the text representation ('OK')
2. A stub for the notebook convert endpoint. Right now it returns static HTML. Later it will return the result of a notebook->html conversion

Commit:
Add notebook conversion via nbconvert

Import the nbconvert library and programmatically perform the request json -> html conversion

I tested this manually using a few notebooks, including tables and images.
output was visually equivalent in the browser to nbconvert on command line, and to Calhoun

Commit:
Update route decorators, names

Update route decorators to shorthand versions
Update handler names to match Calhoun, as these get picked up as default route names

Commit:
Update README.md

Added some information regarding the project description, choice of framework, and background. Will add configuration and deployment once I've worked on those features.

Commit:
Added basic SAM authorization

In Calhoun, the service this is based on, the /api/convert endpoint is authorized via a call to the SAM authorization service that powers Terra. I'm not entirely sure this is necessary- the user is only getting access to a reformated version of data they are sending in. Perhaps checking that the user is registered is a way of protecting from automated DOS attacks. Perhaps it was done to avoid changes to the security boundary for compliance. I will follow up here. For now, I am calling the same authorization service endpoint.

Added basic call to the SAM authorization service.
I don't love the elifs for handling the response out of SAM. Better error handling TBD

Also TBD is parameterizing the SAM root from config.

Updated readme

Commit:
Improved error handling of the authorization call to sam

More effectively handling scenarios where the user is unauthorized, not enabled in Terra, otherwise Forbidden, or there is a problem contacting the SAM authorization service.

Commit:
Loaded SAM_ROOT from config, simplified error handling

Passing in SAM_ROOT from a config file, to facilitate different environments.

Simplified error handling in authorization functions to make use of Sanic's built in exception handling

ericsong commented on README.md:
```diff
@@ -1,2 +1,52 @@
+/status is a health-check endpoint. It accepts GET requests and returns 200 'OK'.
+
+/api/convert is the primary conversion endpoint. It accepts POST requests containing notebook JSON and returns HTML for displaying the notebook's content in the browser.
+This endpoint is authorized via Terra's SAM authorization service. Calls therefore require an auth header supplying a Terra Oauth2 bearer token.
```
does it have to be a `Terra` oauth token?

bradtaylor replied:
I'll clarify the comment. It's an Oauth2 token representing the user's Google identity, either from Google SSO or gcloud auth login. That Google identity must be known to Terra (user must be registered and enabled in SAM)

ericsong commented on main.py:
```diff
@@ -0,0 +1,23 @@
+
+# Webservice routing
+app = Sanic('douglass')
+app.config.from_pyfile('config.py')
```
how is the application going to pick the right config file based on the environment it's running in?

bradtaylor replied:
I'm hoping to get some advice on this from the Terra UI team. They wrote a deploy script that copied one of the environment examples from config.ENV.py to config.py I will probably do something similar

ericsong commented on main.py:
```diff
@@ -0,0 +1,23 @@
+app.config.from_pyfile('config.py')
+
+
+@app.get('/')
```
`/status` for the status endpoint? not sure what Broad convention is

bradtaylor replied:
This should indeed by /status. I will fix

ericsong commented on main.py:
```diff
@@ -0,0 +1,23 @@
+
+
+if __name__ == '__main__':
+    app.run(port=8000)
```
does production use a different port?

bradtaylor replied:
I haven't really touched configuring ports etc for production yet. I'm looking into how Google App Engine deploys work. I'm intending to make an additional PR for production deployments onto GAE

ericsong reviewed (commented):
The application code looks good to me. Most of my comments are around how this gets productionalized. Ex. how it handles the different environments.

Also a note that `Terra` is not the only user of this service so the comments/code should not have that assumption baked into them.

ericsong commented on utils.py:
```diff
@@ -0,0 +1,81 @@
+
+
+# Query Terra's authorization service SAM to determine user authorization status. Return auth status boolean
+async def check_sam_authorization(request, sam_root):
```
Also, I believe you can indicate private methods in python by prefixing method names with `__`

bradtaylor replied:
Nice! I never knew that. I'll update this to make use of it

bradtaylor commented:
@ericsong Thanks for the review! I haven't productionized the deployment yet. My intention is to get some advice from folks on the Terra UI team regarding GAE deploy and then handle that as a follow-on PR.

In speaking of Terra, my intention is to refer to the platform-services layer (the way All of Us talks about 'FireCloud'). Would it be helpful to update this to Terra/FireCloud, in case the latter is how it's referred to in AoU documentation

Commit:
Make authorization functions private

Authorization functions are only meant to be called by the 'authorized' decorator.

Commit:
Updated status endpoint to '/status'

Had the status endpoint at '/'. Moved to '/status' to align with Terra API conventions

Commit:
Updated README.md

Clarified the relationship of this project to the Terra/FireCloud suite of microservices. Made a small update to the auth comment. Removed some language refercing this service's origins as a personal project.

ericsong commented:
Oh ok. I didn't realize Terra referred to the entire platform-services layer. In my mind, I was thinking that it was referring to the user facing app like AoU.

In that case, I think the way the term is used makes sense.

calbach commented on utils.py:
```diff
@@ -0,0 +1,81 @@
+        @wraps(f)
+        async def decorated_function(request, *args, **kwargs):
+            # Check the Terra authorization service SAM for user auth status
+            user_is_authorized = await __check_sam_authorization(request, sam_root)
```
What action is being authorized here? What permission do I need?

Are you just checking that the user is authenticated? (if so, should just rename to is_authenticated)
